### PR TITLE
SEM module updates

### LINKED ISSUE #2443: lavaan ERROR: missing observed variables in dataset
## Template for bug reports

* JASP version: 0.8.6
* OS name and version: Windows 7
* Analysis: SEM using lavaan. Subcommand: lav_data_full
* Steps to reproduce: Use any dataset with variable names that include a '_'

I have a dataset that contains variables that include a '\_'.  
I can succesfully run a simple SEM if I only specify variables without a '\_', such as:  
 `abc =~ a + b + c`.  
The error output tells me that variables are missing in the dataset.  
What is striking, is that the Xbase64 string that it gives me, still contains the '\_'. I assume that this is the issue.

NOTE: all letters are blue in the model syntax box in JASP, except after the underscore, they turn black.

### PR DESCRIPTION
**Pull request by vankesteren**
### I would like quite some checks before this pull request is merged:

- [x] @vandenman can you try to run arbitrary `R` code by hacking the SEM syntax editor? Hint: check the `checkSyntax` method in the file `semsimpleform.cpp`.
- [x] @vandenman can you in general see if some of the errors you reported earlier are fixed now?
- [x] @boutinb can you check the edits in the `C++` code I've made, especially to the `boundtextedit` and the `textmodellavaan`? 
- [x] @JorisGoosen can you make general comments and see if nothing bad was done to your filter functionality?

### On each build, could you check whether the fonts look pretty like this in the syntax editor?
- [x] Windows (@vandenman)
- [x] Mac (@boutinb)
- [x] Linux (@JorisGoosen or @akashrajkn once your environment is set up ;) 

image

Here is the model that works with the political democracy example dataset
```
# Bollen's political democracy model

# measurement model
ind60 =~ x1 + x2 + x3
dem60 =~ y1 + y2 + y3 + y4
dem65 =~ y5 + y6 + y7 + y8

# regressions
dem60 ~ ind60
dem65 ~ ind60 + dem60 

# residual correlations
y1 ~~ y5
y2 ~~ y4 + y6
y3 ~~ y7
y4 ~~ y8
y6 ~~ y8
```

### After these are checked off, this pull request may be merged.

fixes #2443 and several internally documented SEM module errors.

### CODE REVIEW

**Comment by vankesteren** [comment]
This fails because of the ancova bayesian test, just like all the other branches...

**Comment by vandenman** [comment]
- **Arbitrary code can crash JASP.** Using a \ in the model leads to a crash.
- The previously reported error are all fixed. 
- The font looks pretty on windows!

I still have to take an in depth look at the code (and regular expressions) but during some testing everything looked fine.
[semSpecial2018.zip](https://github.com/jasp-stats/jasp-desktop/files/1918609/semSpecial2018.zip)

**Comment by vankesteren** [comment]
Oh great, thanks! I thought I had caught that, but that should be an easy fix. Will fix on wednesday

**Comment by akashrajkn** [comment]
Tested on my system. Looks good

**Comment by vandenman** [comment]
Here another set of bugs. Most of these are edge cases, although it would be nice if people could use something like α (and perhaps even ✈). 
[semBugs2.zip](https://github.com/jasp-stats/jasp-desktop/files/1946889/semBugs2.zip)

File: JASP-Desktop/enginesync.cpp [hunk 1]
```diff
@@ -417,6 +429,8 @@ void EngineSync::runScriptOnProcess(RScriptStore * scriptStore, int processNo)
 
 void EngineSync::sendMessages()
 {	
+	std::cout << "void EngineSync::sendMessages()\n" << std::flush;
```
Comment (JorisGoosen) [inline JASP-Desktop/enginesync.cpp]: Maybe take out the debugtraces?

File: JASP-Desktop/main.cpp [hunk 2]
```diff
@@ -32,7 +32,7 @@ int main(int argc, char *argv[])
 	// Follow status of https://bugreports.qt.io/browse/QTBUG-61430 for 
 	// future permanent fix.
 	// This does slow down QML quite a bit and disables gradients
-	qputenv("QT_QUICK_BACKEND", "software");
```
Comment (JorisGoosen) [inline JASP-Desktop/main.cpp]: Should probably leave this turned on till the bug QTBUG-61430 is fixed?

File: JASP-Desktop/widgets/boundtextedit.cpp [hunk 3]
```diff
@@ -30,21 +31,29 @@ BoundTextEdit::BoundTextEdit(QWidget *parent) :
 	_model = new TextModelLavaan(this);
 	setDocument(_model);
 
-	connect(this, SIGNAL(cursorPositionChanged()), this, SLOT(cursorPositionChangedHandler()));
-	connect(_model, SIGNAL(errorStateChanged()), this, SLOT(errorStateChangedHandler()));
-	connect(_model, SIGNAL(contentsChanged()), this, SLOT(contentsChangedHandler()));
+	connect(this,	&BoundTextEdit::cursorPositionChanged,	this, &BoundTextEdit::cursorPositionChangedHandler);
```
Comment (JorisGoosen) [inline JASP-Desktop/widgets/boundtextedit.cpp]: Nice! :)

File: JASP-Engine/engine.h [hunk 4]
```diff
@@ -93,8 +93,11 @@ class Engine
 
 	int _slaveNo = 0;
 
-	bool filterChanged = false;
-	std::string filter = "return(genFilter)", generatedFilter = "";
+	bool _filterChanged = false;
```
Comment (JorisGoosen) [inline JASP-Engine/engine.h]: Goed idee

**Comment by vankesteren** [comment]
@vandenman nice finds! Indeed, they are quite edgy cases, all using the backticks. I feel that the way JASP responds to those issues is fine; backticks are not really meant to be used in this module anyway, and I doubt people will use them extensively in the ways you did. In any case, this does not really pose an issue for security or anything like that. 

Just don't call your variables `(╯°□°）╯︵ ┻━┻)` 😄 

The disappearing labels of unicode characters such as Ω issue is perhaps a little more problematic. I'll look into this now.

File: JASP-Engine/JASP/R/semsimple.R [hunk 5]
```diff
@@ -87,528 +87,548 @@
...
+
+  if (perform == "run" && options$Data != "varcov") {
+    dataset <- .readDataSetToEnd(columns = variables)
+  } else if (perform == "run" && options$Data == "varcov") {
+    dataset <- .readDataSetToEnd(all.columns = TRUE)
+  } else {
+    dataset <- dheader
+  }
+
+  #browser()
```
Comment (vandenman) [inline JASP-Engine/JASP/R/semsimple.R]: Can be removed?

Comment (vankesteren) [reply JASP-Engine/JASP/R/semsimple.R]: yes, just saw that too :)

**Comment by vankesteren** [comment]
Ready to merge! 👌👌👌

**Pull request merged**